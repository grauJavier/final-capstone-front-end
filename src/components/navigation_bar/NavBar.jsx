import { GiHamburgerMenu } from 'react-icons/gi';
import { AiFillCloseCircle } from 'react-icons/ai';
import { useState, useEffect } from 'react';
import { useLocation, useNavigate } from 'react-router-dom';
import Logout from '../register_login_forms/Logout';
import logoBlack from '../../assets/logo-black.png';

const NavBar = () => {
  const [show, setShow] = useState(false);
  const [showItem, setShowItem] = useState(true);
  const path = useLocation();
  const navigate = useNavigate();
  const [windowWidth, setwindowWidth] = useState(window.innerWidth);

  // Find the right width when the window resize.
  useEffect(() => {
    const handleResize = () => {
      setwindowWidth(window.innerWidth);
    };

    window.addEventListener('resize', handleResize);

    return () => {
      window.removeEventListener('resize', handleResize);
    };
  }, []);

  // Hide the NavBar when tapping outside of the element.
  const hideNav = (event) => {
    const container = document.getElementById('menu-container');
    if (event.target !== container) {
      container.classList.remove('animate-show-menu');
      container.classList.add('animate-hide-menu');
      window.removeEventListener('click', hideNav);
      setTimeout(() => {
        setShow(false);
        setShowItem(true);
      }, 600);
    }
  };

  const showHideNav = (event) => {
    event.stopPropagation();
    if (show) {
      const container = document.getElementById('menu-container');
      container.classList.remove('animate-show-menu');
      container.classList.add('animate-hide-menu');
      setTimeout(() => {
        setShow(false);
        setShowItem(true);
      }, 600);
    } else {
      setShow(true);
      setShowItem(false);
      window.addEventListener('click', hideNav);
    }
  };

  const elements = [
    ['Places', '/places'],
    ['New Reservation', '/reservation'],
    ['My Reservations', '/my-reservations'],
    ['My Places', '/my-places'],
    ['New Place', '/new-place'],
  ];

  return (
    <>
      {showItem && windowWidth < 1024 && (
        <GiHamburgerMenu
          className="fixed ml-[15px] mt-[15px] text-[1.5rem] text-black bg-white p-2 w-12 h-12 rounded-full drop-shadow-xl"
          onClick={showHideNav}
        />
      )}
      {(show || windowWidth >= 1024) && (
        <div
          className="min-h-screen w-3/4 max-w-[320px] flex flex-col primary-font fixed z-50 bg-white drop-shadow-xl animate-show-menu lg:shadow-none lg:border-r-2 lg:border-stone-100 lg:max-w-[15vw] lg:animate-none"
          id="menu-container"
        >
          <AiFillCloseCircle
            className="ml-[15px] mt-[15px] text-[1.5rem] lg:hidden"
            onClick={showHideNav}
          />
          <img src={logoBlack} className="w-[65%] lg:mt-12 mx-auto" />
          <nav className="m-auto ml-[15px] mr-0">
            <ul className="flex flex-col content-center min-h-[70vh] justify-evenly">
              {elements.map((element, index) => (
                <li
                  key={index}
                  className={`py-[10px] pl-[10px] font-bold cursor-pointer ${
                    path.pathname.startsWith(element[1]) ? 'bg-login-green text-white' : ''
                  }`}
                  onClick={() => navigate(`${element[1]}`)}
                >
                  {element[0]}
                </li>
              ))}
            </ul>
          </nav>
          <Logout />
        </div>
      )}
    </>
  );
};

export default NavBar;
