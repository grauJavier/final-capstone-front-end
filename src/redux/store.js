import { configureStore, combineReducers } from '@reduxjs/toolkit';
import userReducer from './user/userSlice';
import { persistStore, persistReducer } from 'redux-persist';
import sessionStorage from 'redux-persist/lib/storage/session'; // defaults to localStorage for web
import { createTransform } from 'redux-persist';
import placesSlice from "./places/placesSlice";
import reservationReducer from './reservation_form/reservationSlice.js';
import myReservationsReducer from './my_reservations/myReservationsSlice.js';
import noticeAlertReducer from './notice_alert/noticeAlertSlice.js';

// Create a personalized transform.
const userTransform = createTransform(
  (inboundState) => {
    return {
      currentUser: inboundState.currentUser,
      token: inboundState.token
    };
  },

  (outboundState) => {
    return {
      ...outboundState
    };
  },
  // Apply this transform only to the user state.
  { whitelist: ['user'] }
);

const persistConfig = {
  key: 'root',
  storage: sessionStorage,
  whitelist: ['user'],
  transforms: [userTransform]
};

const reducer = combineReducers({
  noticeAlert: noticeAlertReducer,
  user: userReducer,
  placesSlice,
  reservation: reservationReducer,
  myReservations: myReservationsReducer,
});

const persistedReducer = persistReducer(persistConfig, reducer);

const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: false
    }),
});

const persistor = persistStore(store);

export { store, persistor };
