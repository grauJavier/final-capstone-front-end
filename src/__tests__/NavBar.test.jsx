import { render, waitFor, fireEvent } from '@testing-library/react';
import NavBar from '../components/navigation_bar/NavBar';
import { describe, it, expect, beforeEach } from 'vitest';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { store } from '../redux/store';

describe('NavBar Component', () => {
  let component;

  beforeEach(() => {
    component = render(
      <Provider store={store}>
        <Router>
          <NavBar />
        </Router>
      </Provider>
    );
  });

  it('NavBar renders correctly', async () => {
    await waitFor(() => {
      expect(NavBar).toMatchSnapshot();
    });
  });

  it('Clicking "Places" link navigates to the "/places" path', async () => {
    const placesLink = component.getByText('Places');
    fireEvent.click(placesLink);

    await waitFor(() => {
      expect(window.location.pathname).toBe('/places');
    });
  });

  it('Clicking "New Reservation" link navigates to the "/reservation" path', async () => {
    const newReservationLink = component.getByText('New Reservation');
    fireEvent.click(newReservationLink);

    await waitFor(() => {
      expect(window.location.pathname).toBe('/reservation');
    });
  });

  it('Clicking "My Reservations" link navigates to the "/my-reservations" path', async () => {
    const myReservationsLink = component.getByText('My Reservations');
    fireEvent.click(myReservationsLink);

    await waitFor(() => {
      expect(window.location.pathname).toBe('/my-reservations');
    });
  });

  it('Clicking "My Places" link navigates to the "/my-places" path', async () => {
    const myPlacesLink = component.getByText('My Places');
    fireEvent.click(myPlacesLink);

    await waitFor(() => {
      expect(window.location.pathname).toBe('/my-places');
    });
  });

  it('Clicking "New Place" link navigates to the "/new-place" path', async () => {
    const newPlaceLink = component.getByText('New Place');
    fireEvent.click(newPlaceLink);

    await waitFor(() => {
      expect(window.location.pathname).toBe('/new-place');
    });
  });
});
