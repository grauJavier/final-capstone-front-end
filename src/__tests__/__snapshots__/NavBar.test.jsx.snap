// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NavBar Component > NavBar renders correctly 1`] = `[Function]`;
